### DISCUSSION TITLE
How do I convert a UIImage to a coil image

### DISCUSSION THREAD

**Discussion by AlexanderEggers** [coil-kt/coil#2229] [Q&A]
I have a use-case in my app where I need to convert a UIImage to a coil image. I'm trying to build a custom convert for using blurhash in my project. Coil doesn't seem to support this at the moment.

**Comment by colinrtwhite** [comment]
I don't think this is possible - `Image`s need to be able to render to a Skia canvas, which I'm not sure an iOS `UIImage` can.

**Comment by kihaki** [comment]
You can convert an UIImage to a ByteArray, [peekaboo](https://github.com/onseok/peekaboo) does this

```
@OptIn(ExperimentalForeignApi::class)
private fun UIImage.toByteArray(compressionQuality: Double): ByteArray {
    val validCompressionQuality = compressionQuality.coerceIn(0.0, 1.0)
    val jpegData = UIImageJPEGRepresentation(this, validCompressionQuality)!!
    return ByteArray(jpegData.length.toInt()).apply {
        memcpy(this.refTo(0), jpegData.bytes, jpegData.length)
    }
}
```

The `ByteArray` can then probably be converted into a `Bitmap` which then can be used in a coil `Image`.

I am currently on the second step (ByteArray -> Bitmap), will post again once I got it working..

Di you already make it work @AlexanderEggers ?

**Comment by kihaki** [comment]
I have managed to use the above code to create an `Buffer` that contains the `UIImage`s content as Jpeg ByteArray

```
val jpegByteArray = uiImage.toByteArray(80.0)
val buffer = Buffer()
buffer.write(jpegByteArray)
```

This probably copies the whole `ByteArray` without any gain, but I have so far not found a zero-copy method of turning a `ByteArray` into a `Buffer` - if someone has something like that, that would be great.

The result can then be passed to a coil `SourceFetchResult`

```
SourceFetchResult(
  source = ImageSource(buffer, FileSystem.SYSTEM),
  mimeType = "image/jpeg",
  dataSource = DataSource.MEMORY
)
```

which then can be used inside of a custom `Fetcher`.

This works for picking an image from the iOS PhotoPicker and display that image in a `LazyColumn` in Compose Multiplatform with Coil. Performance so far is great on the emulator, though to be fair, the emulator runs on an M1, which might just be so powerful as not to notice any performance problems.. haven't tested it on a real iOS device yet.
